From: Benjamin Drung <benjamin.drung@canonical.com>
Date: Tue, 19 Nov 2024 17:50:00 +0100
Subject: test: print all server startup logs in V=1 or V=2 mode

In V=0 mode only print "Waiting for the server to startup" once and do
not print anything from the server startup logs.

In V=1 and V=2 print all lines from the server startup while waiting
for the server to be ready to ease debugging (e.g. when the kernel
crashes the last 10 lines will not be enough).

Forwarded: https://github.com/dracut-ng/dracut-ng/pull/1005
---
 test/test-functions | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/test/test-functions b/test/test-functions
index be35a9d..bffcd50 100644
--- a/test/test-functions
+++ b/test/test-functions
@@ -58,12 +58,18 @@ test_dracut() {
 # Wait for the server QEMU has been started up.
 # It should print "Serving" in the server.log in that case.
 wait_for_server_startup() {
-    local server_pid
+    local lines printed_lines=0 server_pid
     server_pid=$(cat "$TESTDIR"/server.pid)
 
+    echo "Waiting for the server to startup"
     while ! grep -q Serving "$TESTDIR"/server.log; do
-        echo "Waiting for the server to startup"
-        tail "$TESTDIR"/server.log
+        if [ "$V" -ge 1 ]; then
+            lines=$(wc -l "$TESTDIR"/server.log | cut -f 1 -d ' ')
+            if [ "$lines" -gt "$printed_lines" ]; then
+                tail -n "+$((printed_lines + 1))" "$TESTDIR"/server.log
+                printed_lines=$lines
+            fi
+        fi
         if ! test -f "/proc/$server_pid/status"; then
             echo "Error: Server QEMU process $server_pid is gone. Please check $TESTDIR/server.log for failures." >&2
             return 1
